### PR TITLE
Blob Java

### PR DESCRIPTION
**Pull request by kirkeaton**
1. Added foam.blob.Buffer java implementation
2. Added getSize and setSize to Blob interface
3. Added AbstractBlobService in JS and Java
4. Merged Blob_web into Blob.js
5. Added SubBlob Java implementation

### CODE REVIEW

File: src/foam/blob/Blob.js [hunk 1]
```diff
@@ -29,8 +55,22 @@ foam.INTERFACE({
...
         {
-          name: 'offset',
-          of: 'Long'
+          class: 'Int',
+          name: 'offset'
+        }
+      ]
+    },
+    {
+      name: 'getSize',
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: Oh I didn't notice you removed the size property.  If we're going to keep it removed, which is fine, we only need getSize.

Blobs are meant to be immutable, so no need for setSize.

File: src/foam/blob/Blob.js [hunk 2]
```diff
@@ -47,60 +87,32 @@ foam.INTERFACE({
     {
       name: 'put',
       returns: 'Promise',
-      args: [
-        {
-          name: 'blob',
-          of: 'foam.blob.Blob'
-        }
-      ]
+      args: [ 'blob' ]
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: I'd rather keep the additional information unless it's causing problems.  In the future we will use it to remove the need to specify javaType in the args

File: src/foam/blob/Blob.js [hunk 3]
```diff
@@ -154,6 +159,28 @@ foam.CLASS({
...
 foam.CLASS({
   package: 'foam.blob',
+  name: 'AbstractBlobService',
+  abstract: true,
+
+  implements: ['foam.blob.BlobService'],
+
+  methods: [
+    function put(blob) {
+      return this.put_(this.__context__, blob);
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: Add an .in(x) like we do for DAO to return a new ProxyDAO to this in the specified context.

Comment (adamvy) [reply src/foam/blob/Blob.js]: PRoxyBlobService I mean

File: src/foam/blob/Blob.js [hunk 4]
```diff
@@ -175,12 +202,15 @@ foam.CLASS({
   extends: 'foam.blob.AbstractBlob',
   properties: [
     {
+      class: 'foam.core.Blob',
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: Just clasS: 'Blob' is fine.  foam.core is the default

File: src/foam/blob/Blob.js [hunk 5]
```diff
@@ -250,7 +283,7 @@ foam.CLASS({
   ],
   properties: [
     {
-      class: 'String',
+      class: 'Int',
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: Id should be a string not an int, the real BlobServices use a hex representation of the SHA256 hash of the blob.  Only in TestBlobService did we use an int, but that was just for convenient, we can convert the ints to strings in TestBlobService.

File: src/foam/blob/Blob.js [hunk 6]
```diff
@@ -311,7 +344,7 @@ foam.CLASS({
       name: 'fd'
     },
     {
-      class: 'Long',
+      class: 'Int',
```
Comment (adamvy) [inline src/foam/blob/Blob.js]: Why this change?  Blob's can be quite large.

File: src/foam/blob/Blob.js (hunk 1 above)
Comment (kirkeaton) [reply src/foam/blob/Blob.js]: @adamvy The reason for removing it was that the getSize & setSize functions weren't being properly generated for ProxyBlob

Comment (adamvy) [reply src/foam/blob/Blob.js]: Sure that's fine, let's keep it this way, but let's remove setSize() entirely, size is meant to be read only as blobs are immutable.

Comment (kirkeaton) [reply src/foam/blob/Blob.js]: @adamvy Removed the setSize function

File: src/foam/blob/Blob.js (hunk 4 above)
Comment (kirkeaton) [reply src/foam/blob/Blob.js]: Removed the prefix

File: src/foam/blob/Blob.js (hunk 3 above)
Comment (kirkeaton) [reply src/foam/blob/Blob.js]: @adamvy Done.

File: src/foam/blob/Blob.js (hunk 2 above)
Comment (kirkeaton) [reply src/foam/blob/Blob.js]: Okay, I've added it back
